File: src/jvmTest/java/org/snakeyaml/engine/v2/api/LoadSettingsTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.v2.api;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import java.util.Map;

import it.krzeminski.snakeyaml.engine.kmp.api.*;
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings.SpecVersionMutator;
import it.krzeminski.snakeyaml.engine.kmp.exceptions.DuplicateKeyException;
import it.krzeminski.snakeyaml.engine.kmp.schema.CoreSchema;

@Tag("fast")
class LoadSettingsTest {

  @Test
  @DisplayName("Accept only YAML 1.2")
  void acceptOnly12() {
    SpecVersionMutator strict12 = t -> {
      if (t.getMajor() != 1 || t.getMinor() != 2) {
        throw new IllegalArgumentException("Only 1.2 is supported.");
      } else {
        return t;
      }
    };

    LoadSettings settings = LoadSettings.builder().setVersionFunction(strict12).build();
    Load load = new Load(settings);
    try {
      load.loadOne("%YAML 1.1\n...\nfoo");
      fail();
    } catch (IllegalArgumentException e) {
      assertEquals("Only 1.2 is supported.", e.getMessage());
    }
  }

  @Test
  @DisplayName("Do not allow duplicate keys")
  void doNotAllowDuplicateKeys() {
    LoadSettings settings = LoadSettings.builder().setAllowDuplicateKeys(false).build();
    Load load = new Load(settings);
    try {
      load.loadOne("{a: 1, a: 2}");
      fail("Duplicate keys must not be allowed.");
    } catch (DuplicateKeyException e) {
      assertTrue(e.getMessage().contains("found duplicate key a"));
    }
  }

  @Test
  @DisplayName("Do not allow duplicate keys by default")
  void doNotAllowDuplicateKeysByDefault() {
    Load load = new Load();
    try {
      load.loadOne("{a: 1, a: 2}");
      fail("Duplicate keys must not be allowed.");
    } catch (DuplicateKeyException e) {
      assertTrue(e.getMessage().contains("found duplicate key a"));
    }
  }

  @Test
  @DisplayName("Allow duplicate keys")
  void allowDuplicateKeysWhenSpecified() {
    LoadSettings settings = LoadSettings.builder().setAllowDuplicateKeys(true).build();
    Load load = new Load(settings);
    Map<String, Integer> map = (Map<String, Integer>) load.loadOne("{a: 1, a: 2}");
    assertEquals(Integer.valueOf(2), map.get("a"));
  }

  @Test
  @DisplayName("Set and get custom property")
  void customProperty() {
    SomeKey key = new SomeKey();
    LoadSettings settings = LoadSettings.builder().setCustomProperty(key, "foo")
        .setCustomProperty(SomeStatus.DELIVERED, "bar").build();
    assertEquals("foo", settings.getCustomProperty(key));
    assertEquals("bar", settings.getCustomProperty(SomeStatus.DELIVERED));
  }

  @Test
  @DisplayName("Set and get custom I/O buffer size")
  void bufferSize() {
    LoadSettings settings = LoadSettings.builder().setBufferSize(4096).build();
    assertEquals(Integer.valueOf(4096), settings.bufferSize);
  }

  @Test
  @DisplayName("Use Core schema by default")
  void defaultSchema() {
    LoadSettings settings = LoadSettings.builder().build();
    assertEquals(CoreSchema.class, settings.schema.getClass());
  }

  public enum SomeStatus implements SettingKey {
    ORDERED, DELIVERED
  }

  public static final class SomeKey implements SettingKey {

  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/api/LoadSettingsTest.kt
@file:Suppress("UNCHECKED_CAST")
package it.krzeminski.snakeyaml.engine.kmp.api

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.assertions.throwables.shouldThrowWithMessage
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldContain
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings.SpecVersionMutator
import it.krzeminski.snakeyaml.engine.kmp.exceptions.DuplicateKeyException
import it.krzeminski.snakeyaml.engine.kmp.schema.CoreSchema

class LoadSettingsTest : FunSpec({
    test("accept only YAML 1.2") {
        val strict12 = SpecVersionMutator { t ->
            if (t.major != 1 || t.minor != 2) {
                throw IllegalArgumentException("Only 1.2 is supported.")
            } else {
                t
            }
        }

        val settings = LoadSettings(versionFunction=strict12)
        val load = Load(settings)
        shouldThrowWithMessage<IllegalArgumentException>(message = "Only 1.2 is supported.") {
            load.loadOne("%YAML 1.1\n...\nfoo")
        }
    }

    test("do not allow duplicate keys") {
        val settings = LoadSettings(allowDuplicateKeys=false)
        val load = Load(settings)
        shouldThrow<DuplicateKeyException> {
            load.loadOne("{a: 1, a: 2}")
        }.also {
            it.message shouldContain "found duplicate key a"
        }
    }

    test("do not allow duplicate keys by default") {
        val load = Load()
        shouldThrow<DuplicateKeyException> {
            load.loadOne("{a: 1, a: 2}")
        }.also {
            it.message shouldContain "found duplicate key a"
        }
    }

    test("allow duplicate keys") {
        val settings = LoadSettings(allowDuplicateKeys=true)
        val load = Load(settings)
        val map = load.loadOne("{a: 1, a: 2}") as Map<String, Int>
        map["a"] shouldBe 2
    }

    test("set and get custom property") {
        val key = SomeKey()
        val settings = LoadSettings(
            customProperties = mapOf(
                key to "foo",
                SomeStatus.DELIVERED to "bar",
            ),
        )
        settings.customProperties[key] shouldBe "foo"
        settings.customProperties[SomeStatus.DELIVERED] shouldBe "bar"
    }

    test("set and get custom I/O buffer size") {
        val settings = LoadSettings(bufferSize=4096)
        settings.bufferSize shouldBe 4096
    }

    test("use Core schema by default") {
        val settings = LoadSettings()
        settings.schema::class shouldBe CoreSchema::class
    }

    test("copy DSL smoke test") {
        val settings1 = LoadSettings()
        val settings2 = settings1.copy {
            label = "foobarbaz"
        }
        settings1.label shouldNotBe settings2.label
    }
})

private enum class SomeStatus : SettingKey {
    ORDERED, DELIVERED
}

private class SomeKey : SettingKey
